include/kernels/restrictor.cuh

```diff
@@ -2,211 +2,166 @@
-
-        int x = parity*arg.in.VolumeCB() + x_cb;
-        int x_coarse = arg.fine_to_coarse[x];
-        int parity_coarse = (x_coarse >= arg.out.VolumeCB()) ? 1 : 0;
-        int x_coarse_cb = x_coarse - parity_coarse*arg.out.VolumeCB();
-
-        for (int coarse_color_block=0; coarse_color_block<coarseColor; coarse_color_block+=coarse_colors_per_thread) {
-          complex<Float> tmp[fineSpin*coarse_colors_per_thread];
-          rotateCoarseColor<Float,fineSpin,fineColor,coarseColor,coarse_colors_per_thread>
-            (tmp, arg.in, arg.V, parity, arg.nParity, x_cb, coarse_color_block);
-
-          for (int s=0; s<fineSpin; s++) {
-            for (int coarse_color_local=0; coarse_color_local<coarse_colors_per_thread; coarse_color_local++) {
-              int c = coarse_color_block + coarse_color_local;
-              arg.out(parity_coarse,x_coarse_cb,arg.spin_map(s,parity),c) += tmp[s*coarse_colors_per_thread+coarse_color_local];
-            }
+  template <typename Arg> struct Restrictor {
+    static constexpr unsigned block_size = Arg::block_size;
+    static constexpr int coarse_color_per_thread = coarse_colors_per_thread<Arg::fineColor, Arg::coarseColor>();
+    using vector = vector_type<complex<typename Arg::real>, Arg::coarseSpin*coarse_color_per_thread>;
+    const Arg &arg;
+    constexpr Restrictor(const Arg &arg) : arg(arg) {}
+    static constexpr const char *filename() { return KERNEL_FILE; }
+
+    __device__ __host__ inline void operator()(dim3 block, dim3 thread)
+    {
+      int x_coarse = block.x;
+      int x_fine_offset = thread.x;
+      int coarse_color_thread = block.y * arg.block_dim.y + thread.y;
+
+      vector reduced;
+      if (x_fine_offset < arg.aggregate_size) {
+        // all threads with x_fine_offset greater than aggregate_size_cb are second parity
+        const int parity_offset = x_fine_offset >= arg.aggregate_size_cb ? 1 : 0;
+        const int x_fine_cb_offset = x_fine_offset - parity_offset * arg.aggregate_size_cb;
+        const int parity = arg.nParity == 2 ? parity_offset : arg.parity;
+
+        // look-up map is ordered as (coarse-block-id + fine-point-id),
+        // with fine-point-id parity ordered
+        const int x_fine_site_id = (x_coarse * 2 + parity) * arg.aggregate_size_cb + x_fine_cb_offset;
+        const int x_fine = arg.coarse_to_fine[x_fine_site_id];
+        const int x_fine_cb = x_fine - parity * arg.in.VolumeCB();
+
+        const int coarse_color_block = coarse_color_thread * coarse_color_per_thread;
+
+        vector_type<complex<typename Arg::real>, Arg::fineSpin*coarse_color_per_thread> tmp;
+        rotateCoarseColor(tmp, arg, parity, x_fine_cb, coarse_color_block);
+
+        // perform any local spin coarsening
+#pragma unroll
+        for (int s=0; s<Arg::fineSpin; s++) {
+#pragma unroll
+          for (int v=0; v<coarse_color_per_thread; v++) {
+            reduced[arg.spin_map(s,parity)*coarse_color_per_thread+v] += tmp[s*coarse_color_per_thread+v];
```

Comment:
`cmac`

Reply:
`cmac` isn't applicable here, there's no multiply here, it's just summation.